Add speed step dropdown to throttles

trainman419 opened the pull request:
Add a speed step dropdown to the extended throttle; this makes it easier to select the speed step mode and makes it possible to select the less common speed step modes exposed by some systems. I've also consolidated the throttle layout code, centralized the logic for determining if throttle icons should be shown, and added unit tests for the control panel layout and some preferences.

Slider:
![extended_slider](https://user-images.githubusercontent.com/103991/76286116-84a3da00-625e-11ea-9044-385aec3b924a.png)

Shunting:
![extended_shunting](https://user-images.githubusercontent.com/103991/76286113-84a3da00-625e-11ea-9e30-c28e39892b2b.png)

Step display (previously speed step):
![extended_step](https://user-images.githubusercontent.com/103991/76286117-853c7080-625e-11ea-95d5-963113e7c644.png)

rhwood commented on java/src/jmri/SpeedStepMode.java:
```diff
@@ -20,30 +20,38 @@
 
-    SpeedStepMode(String name, int numSteps) {
-        this(name, numSteps, 1.0f / numSteps);
+    SpeedStepMode(String name, int numSteps, String peopleKey) {
```
Please change references to `peopleKey` or `peopleName` to `description`, since that is a common term for this kind of key elsewhere in our code.

trainman419 replied:
Done.

danielb987 replied:
@trainman419 
The code still has `peopleKey` instead of `description`. Please change this.

trainman419 replied:
Ooops. Fixed.

danielb987 replied:
Thanks.

devel-bobm commented:
@trainman419, you wrote "...and makes it possible to select the less common speed step modes exposed by some systems. " and "Moving the speed step names from the throttle to java/src/jmri/NamedBeanBundle.properties and attaching them to the speed step mode."

It see two issues with this:

- If the speed-step modes are "system-specific", then the names of the modes ought not be stored in a common .properties file.  Instead, the system's list of supported speed step names ought to be handled in a system-specific way, within  system-specific code.
 
- You have moved the speed-step selector from the "address selection" portion to the "speed/direction" portion of the throttle window.  This implies that "speed step" may be changed "on-the-fly".  At least on LocoNet-based systems, speed-step selection is a function of acquistion.  If changed on-the-fly, it may be necessary to release then re-acquire the loco (i am not sure), but it will be necessary for LocoNet-based systems to send and process LocoNet messages which look a great deal like the ones used for acquiring a loco.  So, for LocoNet, I would think this GUI element makes more sense in the "Address panel", and, once a loco is acquired, the setting should be disabled so that it cannot be changed on-the-fly.

bobjacobsen commented:
> If the speed-step modes are "system-specific", then the names of the modes ought not be stored in a common .properties file. Instead, the system's list of supported speed step names ought to be handled in a system-specific way, within system-specific code.

Not a problem per se.  These keys are used by the `jmri.SpeedStepMode` class that's already at the top level.

pabender commented:
Sent from my iPad

> On Mar 10, 2020, at 8:10 AM, Bob Milhaupt <notifications@github.com> wrote:
> You have moved the speed-step selector from the "address selection" portion to the "speed/direction" portion of the throttle window.
> 

Speed step mode selection has ALWAYS been on the speed/direction portion of the throttle.  Previously it was radio buttons on the speed step mode spinner pane, which was hidden by default.  This change makes the selection more visible.

> At least on LocoNet-based systems, speed-step selection is a function of acquistion.

Speed step mode is changed on a LocoNet system by status editing the slot.  Last time I looked, you could status edit a slot without releasing and reacquiring the slot.  

If the JMRI LocoNet throttle doesn’t handle changing speed steps when it triggers changing the change of speed steps, that sounds like the JMRI LocoNet throttle has a bug in it that needs to be fixed.

Paul

bobjacobsen commented:
One static analysis issue:
```
1. ERROR in /home/travis/build/JMRI/JMRI/java/src/jmri/jmrit/throttle/ThrottlesPreferencesPane.java (at line 245)
796	cbUseFunctionIcon.setSelected(tp.isUsingFunctionIcon());
797	                                 ^^^^^^^^^^^^^^^^^^^^^
798The method isUsingFunctionIcon() from the type ThrottlesPreferences is deprecated
```

The deprecation is new in this PR, with comment "This function is deprecated; use isUsingIcons() to check if icons should be shown instead."

trainman419 commented:
Thanks! I've removed the deprecation notice; it seems like that function is used in a number of places and I think I should leave it as is.

I've updated the remaining properties files and I've done my best to copy over the existing speed step modes and translate the new ones with google translate.

trainman419 commented:
I think this is ready to merge; is there anything else I need to do?

trainman419 commented:
Thanks!
